pkg/cluster/rook_ceph.go

```diff
@@ -196,81 +227,102 @@ func (c *Controller) SetDeviceHealthMetricsReplication(rookVersion semver.Versio
 		minSize = 2
 	}
 
-	c.Log.Debugf("Ensuring %s replication level is %d", CephDeviceHealthMetricsPool, level)
+	poolName := CephDeviceHealthMetricsPool
+	if cephVersion != "" {
+		cephVersionSemver, err := semver.Parse(cephVersion)
+		if err != nil {
+			return false, errors.Wrapf(err, "parse ceph version %q", cephVersion)
+		}
+		if cephVersionSemver.Major >= 17 {
+			poolName = CephDeviceHealthMetricsPoolQuincy
+		}
+	}
+
+	c.Log.Debugf("Ensuring %s replication level is %d", poolName, level)
 
-	err := c.cephOSDPoolSetSize(rookVersion, CephDeviceHealthMetricsPool, level, cephcluster)
+	err := c.cephOSDPoolSetSize(rookVersion, poolName, level, cephcluster)
 	if err != nil {
-		return false, errors.Wrapf(err, "scale %s pool size", CephDeviceHealthMetricsPool)
+		return false, errors.Wrapf(err, "scale %s pool size", poolName)
 	}
-	err = c.cephOSDPoolSetMinSize(rookVersion, CephDeviceHealthMetricsPool, minSize)
+	err = c.cephOSDPoolSetMinSize(rookVersion, poolName, minSize)
 	if err != nil {
-		return false, errors.Wrapf(err, "scale %s pool min_size", CephDeviceHealthMetricsPool)
+		return false, errors.Wrapf(err, "scale %s pool min_size", poolName)
 	}
 
 	return true, nil
 }
 
-func (c *Controller) ReconcileMonCount(ctx context.Context, count int) error {
+func (c *Controller) ReconcileMonCount(ctx context.Context, nodeCount int) error {
 	// single mon for 1 or 2 node cluster, 3 mons for all other clusters
-	if count < 3 {
-		count = 1
-	} else {
-		count = 3
+	monCount := 3
+	if nodeCount < 3 {
+		monCount = 1
 	}
 
 	cluster, err := c.GetCephCluster(ctx)
 	if err != nil {
 		return errors.Wrapf(err, "get CephCluster config")
 	}
 
-	if cluster.Spec.Mon.Count == count {
+	if cluster.Spec.Mon.Count == monCount {
 		return nil
 	}
-	if cluster.Spec.Mon.Count > count {
-		c.Log.Debugf("Will not reduce mon count from %s to %s", cluster.Spec.Mon.Count, count)
+	if cluster.Spec.Mon.Count > monCount {
+		c.Log.Debugf("Will not reduce mon count from %s to %s", cluster.Spec.Mon.Count, monCount)
 		return nil
 	}
 
-	c.Log.Infof("Increasing mon count from %d to %d", cluster.Spec.Mon.Count, count)
-	cluster.Spec.Mon.Count = count
-	_, err = c.Config.CephV1.CephClusters("rook-ceph").Update(ctx, cluster, metav1.UpdateOptions{})
+	c.Log.Infof("Increasing mon count from %d to %d", cluster.Spec.Mon.Count, monCount)
+
+	patches := []k8s.JSONPatchOperation{{
+		Op:    k8s.JSONPatchOpReplace,
+		Path:  "/spec/mon/count",
+		Value: monCount,
+	}}
+
+	_, err = c.JSONPatchCephCluster(ctx, patches)
 	if err != nil {
-		return errors.Wrap(err, "update CephCluster with new mon count")
+		return errors.Wrap(err, "patch CephCluster with new mon count")
 	}
 
 	return nil
 }
 
-func (c *Controller) ReconcileMgrCount(ctx context.Context, rookVersion semver.Version, count int) error {
+func (c *Controller) ReconcileMgrCount(ctx context.Context, rookVersion semver.Version, nodeCount int) error {
```

Comment:
Not a blocker since all implementations seems to follow the same approach, however, why are we not using `req ctrl.Request (ctrl.Result, error)` ?

Also, why the name is Controller instead of RockController for example? 
Also, why we have controllers under the pkg directory? We should not allow export the controllers and inside of the pkg ideally we should have only APIs and code implementations that we want make available for other projects if/when them import this one. Are we importing ecko in any other place as a dependency?

Reply:
Can you create a story in Replicated's internal Shortcut?